### PR TITLE
Fix cache functions inlining

### PR DESCRIPTION
**Pull request by carlocaione**
The current state of the cache framework doesn't allow for the full inlining of the cache functions. Try to fix this problem reworking the header files and move the ARM64 implementation to a fully inlined version by adding a new arch-specific cache-related header file. All the other architectures must do the same if they want to fully inline the cache operations.

### CODE REVIEW

**Review by nashif — DISMISSED** [review]

File: include/zephyr/cache.h [hunk 1]
```diff
@@ -341,8 +257,10 @@ static inline int sys_cache_instr_flush_range(void *addr, size_t size)
  * @retval -ENOTSUP If not supported.
  * @retval -errno Negative errno for other failures.
  */
-__syscall int sys_cache_data_invd_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_invd_range(void *addr, size_t size)
+__syscall int sys_cache_data_invd_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
Comment (stephanosio) [inline include/zephyr/cache.h:261]: Should use `ALWAYS_INLINE` for portability.

File: include/zephyr/cache.h [hunk 2]
```diff
@@ -388,8 +306,10 @@ static inline int sys_cache_instr_invd_range(void *addr, size_t size)
  * @retval -ENOTSUP If not supported.
  * @retval -errno Negative errno for other failures.
  */
-__syscall int sys_cache_data_flush_and_invd_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_flush_and_invd_range(void *addr, size_t size)
+__syscall int sys_cache_data_flush_and_invd_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
Comment (stephanosio) [inline include/zephyr/cache.h:310]: same here

File: include/zephyr/cache.h [hunk 3]
```diff
@@ -294,8 +208,10 @@ static inline int sys_cache_instr_flush_and_invd_all(void)
  * @retval -ENOTSUP If not supported.
  * @retval -errno Negative errno for other failures.
  */
-__syscall int sys_cache_data_flush_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_flush_range(void *addr, size_t size)
+__syscall int sys_cache_data_flush_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
Comment (stephanosio) [inline include/zephyr/cache.h:212]: ditto

File: include/zephyr/cache.h (hunk 1 above)
Comment (carlocaione) [reply include/zephyr/cache.h:261]: Yes, but that would confuse `scripts/build/parse_syscalls.py` that would be unable to parse anything (this is why the attribute is also put at the end). So, either we live with this or we have to modify the parsing script to account for `ALWAYS_INLINE`.

File: include/zephyr/cache.h (hunk 2 above)
Comment (carlocaione) [reply include/zephyr/cache.h:310]: ditto

File: include/zephyr/cache.h (hunk 3 above)
Comment (carlocaione) [reply include/zephyr/cache.h:212]: ditto

File: include/zephyr/cache.h (hunk 1 above)
Comment (gmarull) [reply include/zephyr/cache.h:261]: Sounds like a `parse_syscalls.py` issue that should be resolved.

File: include/zephyr/arch/cache.h [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+#define ZEPHYR_INCLUDE_ARCH_CACHE_H_
+
+/**
+ * @brief Cache Controller Interface
+ * @defgroup cache_arch_interface Cache Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#if defined(CONFIG_ARM64)
```
Comment (povergoing) [inline include/zephyr/arch/cache.h:22]: Might there be too many `#if deifned()` for different `arch` in the future?

Comment (carlocaione) [reply include/zephyr/arch/cache.h:22]: well, it's a common pattern in Zephyr for arch headers: https://github.com/zephyrproject-rtos/zephyr/blob/c4565ad5b6527dd2474067230f98108518324ab4/include/zephyr/arch/cpu.h#L14-L36

Comment (povergoing) [reply include/zephyr/arch/cache.h:22]: Ah, right, forgot it's a header

File: include/zephyr/cache.h (hunk 1 above)
Comment (carlocaione) [reply include/zephyr/cache.h]: done, but be aware that my python-foo is not strong :D

**Review by povergoing — DISMISSED** [review]

File: scripts/build/parse_syscalls.py [hunk 5]
```diff
@@ -83,6 +83,15 @@ def analyze_headers(multiple_directories):
...
                         raise
 
+                # Remove ALWAYS_INLINE to correctly parse fully-inlined
+                # syscalls defined as:
+                #
+                #      __syscall ALWAYS_INLINE sc(...)
+                #
+                # like it is done for some cache functions in
+                # include/zephyr/cache.h
+                contents = re.sub(r'ALWAYS_INLINE', '', contents)
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply scripts/build/parse_syscalls.py:93]: yup, apparently also my regex-foo is not great :coffin:

still the CI is complaining about the `__syscall ALWAYS_INLINE` with `error: duplicate 'inline' declaration specifier `

[the remark this answers is not in the document]
Comment (carlocaione) [reply scripts/build/parse_syscalls.py:93]: @stephanosio what about adding a new `__syscall_force_inline` instead?

[the remark this answers is not in the document]
Comment (stephanosio) [reply scripts/build/parse_syscalls.py:93]: > @stephanosio what about adding a new `__syscall_force_inline` instead?

Sounds like a reasonable idea. Maybe `__syscall_always_inline` to align with the existing nomenclature.

File: include/zephyr/toolchain/common.h [hunk 6]
```diff
@@ -143,8 +143,10 @@
  */
 #ifndef ZTEST_UNITTEST
 #define __syscall static inline
+#define __syscall_always_inline static inline __attribute__((always_inline))
 #else
 #define __syscall
+#define __syscall_always_inline
 #endif /* ZTEST_UNITTEST */
```
Comment (stephanosio) [inline include/zephyr/toolchain/common.h:150]: These changes (including to `parse_syscalls.py`) should go into a separate commit.

[the remark this answers is not in the document]
Comment (carlocaione) [reply include/zephyr/toolchain/common.h:150]: No, because I think this comment is also valid also for `__syscall_always_inline` https://github.com/zephyrproject-rtos/zephyr/blob/2236e31f6a49f9fb4f64cc1ebace010f1c0d86a9/include/zephyr/toolchain/common.h#L138-L142

File: scripts/build/parse_syscalls.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlocaione) [reply scripts/build/parse_syscalls.py]: added

**Comment by carlocaione** [comment]
> whats up with docs failing? is this related to this pr?

no idea. I'm investigating.

**Review by stephanosio — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: include/zephyr/toolchain/common.h [hunk 7]
```diff
@@ -143,8 +143,10 @@
  */
 #ifndef ZTEST_UNITTEST
 #define __syscall static inline
+#define __syscall_always_inline static inline __attribute__((always_inline))
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply include/zephyr/toolchain/common.h:146]: Nope, see https://github.com/zephyrproject-rtos/zephyr/pull/57021#discussion_r1172387821

[the remark this answers is not in the document]
Comment (ceolin) [reply include/zephyr/toolchain/common.h:146]: That is not what I'm saying, I meant `#define __syscall_always_inline ALWAYS_INLINE`

**Review by ceolin — APPROVED** [review]
Looks much cleaner now :)
